src/main/java/org/jboss/pnc/reqour/common/executor/task/TaskExecutorImpl.java

```diff
@@ -45,12 +46,18 @@ public TaskExecutorImpl(ManagedExecutor executor) {
     }
 
     @Override
-    public <REQ, RES> void executeAsync(
+    public <T, R> void executeAsync(
             String taskID,
-            REQ request,
-            Function<REQ, RES> syncExecutor,
-            BiConsumer<RES, Throwable> callbackSender) {
+            Request request,
+            T req,
+            Function<T, R> syncExecutor,
+            TriFunction<T, R, Throwable, R> errorHandler,
+            BiConsumer<Request, R> callbackSender) {
         log.debug("Starting the task with ID={}", taskID);
-        runningTasks.put(taskID, executor.supplyAsync(() -> syncExecutor.apply(request)).whenComplete(callbackSender));
+        runningTasks.put(
+                taskID,
+                executor.supplyAsync(() -> syncExecutor.apply(req))
+                        .handle((res, t) -> errorHandler.apply(req, res, t))
```

Comment:
instead of handle, you could  maybe use exceptionally

Reply:
nice catch! that is more intuitive and we get rid of that custom `TriFunction` :smile: